src/main/java/com/microsoft/sqlserver/jdbc/SQLServerTimeoutManager.java

```diff
@@ -0,0 +1,154 @@
+    private static Map<UUID, List<TimeoutCommand<?>>> timeoutCommands = new HashMap<>();
+    private final static Logger logger = Logger.getLogger("com.microsoft.sqlserver.jdbc.SQLServerTimeoutManager");
+
+    static void startTimeoutCommand(TimeoutCommand<?> timeoutCommand) {
+        if (scheduledTimeoutTasks.isShutdown()) {
+            // reset id counter for timeout commands
+            TimeoutCommand.uniqueId.set(0);
+            scheduledTimeoutTasks = Executors.newScheduledThreadPool(1);
+            timeoutTaskWorker = Executors.newSingleThreadExecutor();
+        }
+
+        Runnable timeoutTaskRunnable = () -> {
+            Future<?> timeoutTask = timeoutTaskWorker.submit(timeoutCommand::interrupt);
+            try {
+                // if the timeout command takes too long to interrupt, cancel it and release
+                timeoutTask.get(10, TimeUnit.SECONDS);
+            } catch (TimeoutException te) {
+                logger.log(Level.WARNING, "Timeout task too long, aborting", te);
+            } catch (InterruptedException | ExecutionException e) {
+                logger.log(Level.FINE, "Unexpected Exception occured in timeout thread.", e);
+            } finally {
+                releaseTimeoutCommand(timeoutCommand);
+            }
+        };
+
+        timeoutCommand.setTimeoutTask(
+                scheduledTimeoutTasks.schedule(timeoutTaskRunnable, timeoutCommand.getTimeout(), TimeUnit.SECONDS));
+        addTimeoutCommand(timeoutCommand);
+    }
+
+    static void releaseAndRemoveTimeoutCommand(TimeoutCommand<?> timeoutCommand) {
+        releaseTimeoutCommand(timeoutCommand);
+        removeTimeoutCommand(timeoutCommand);
+        checkForTimeoutThreadsShutdown();
+    }
+
+    static void releaseTimeoutCommands(UUID connectionId) {
+        synchronized (timeoutCommands) {
+            List<TimeoutCommand<?>> timeouts = timeoutCommands.get(connectionId);
+            if (timeouts != null) {
+                Iterator<TimeoutCommand<?>> iterator = timeouts.iterator();
+                while (iterator.hasNext()) {
+                    TimeoutCommand<?> timeoutCommand = iterator.next();
+                    iterator.remove();
+                    releaseTimeoutCommand(timeoutCommand);
+                }
+
+                checkIfConnectionIsValid(connectionId, timeouts);
+                checkForTimeoutThreadsShutdown();
+            }
+        }
+    }
+
+    private static void checkIfConnectionIsValid(UUID connectionId, List<TimeoutCommand<?>> timeouts) {
+        if (timeouts.isEmpty()) {
+            // if there are no more timeouts, remove the connection from cache
+            timeoutCommands.remove(connectionId);
+        }
+    }
+
+    private static void checkForTimeoutThreadsShutdown() {
+        if (!areTimeoutCommandsAvailable()) {
+            scheduledTimeoutTasks.shutdownNow();
+            timeoutTaskWorker.shutdownNow();

```

Comment:
This change doesn't really accurately resolve the original problem.
>When all connections are closed there should not be any extra threads running and all classes should be unloaded.

This change instead stops all the threads when there are no active commands, which might have some unintended consequences. It's perfectly normal to do work on a single connection. With the current change, threads would be spun up and stopped in between each new query assuming the user consumes all their resources.

Reply:
In the latest commit I modified shutdown logic to only get called when connections are cleaned up and add a test case validating the threads are still around during a single connections lifespan